plasmapy/particles/decorators.py

```diff
@@ -20,488 +18,803 @@
+            raise InvalidParticleError(
+                f"{self.callable_.__name__} does not accept CustomParticle "
+                f"or CustomParticle-like inputs."
+            )
 
-        _charge_number = particle._attributes["charge number"]
+        if not self.allow_particle_lists and isinstance(particle, ParticleList):
+            raise InvalidParticleError(
+                f"{self.callable_.__name__} does not accept ParticleList "
+                "or particle-list-like inputs."
+            )
 
-        must_be_charged = "charged" in require
-        must_have_charge_info = set(any_of) == {"charged", "uncharged"}
+        if not self.allow_custom_particles and isinstance(particle, ParticleList):
+            if any(particle.is_category("custom")):
+                raise InvalidParticleError(
+                    f"{self.callable_.__name__} does not accept CustomParticle "
+                    f"or CustomParticle-like inputs."
+                )
 
-        uncharged = _charge_number == 0
-        lacks_charge_info = _charge_number is None
+    def process_argument(
+        self,
+        parameter: str,
+        argument: Any,
+        Z: Optional[Integral],
+        mass_numb: Optional[Integral],
+    ) -> Any:
+        """
+        Process an argument that has an appropriate annotation.
 
-        if must_be_charged and (uncharged or must_have_charge_info):
-            raise ChargeError(f"A charged particle is required for {funcname}.")
+        Parameters
+        ----------
+        parameter : str
+            The name of the |parameter| that was decorated.
 
-        if must_have_charge_info and lacks_charge_info:
-            raise ChargeError(f"Charge information is required for {funcname}.")
+        argument : object
+            The value of the |argument| associated with ``parameter``.
+
+        Z : integer, optional
+            The |charge number| of an ion or neutral particle.
+
+        mass_numb : integer, optional
+            The mass number of an isotope.
+        """
+        annotation = self.annotations.get(parameter, None)
+
+        if annotation not in _allowed_annotations:
+            return argument
+
+        if annotation in _optional_allowed_annotations and argument is None:
+            return argument
+
+        # This does not yet include cases like Optional[ParticleList],
+        # Union[ParticleList, ParticleLike], etc. and thus needs updating.
+
+        if annotation == (Particle, Particle):  # deprecated
+            if not hasattr(argument, "__len__") or len(argument) != 2:
+                raise ValueError(f"The length of {argument} must be 2.")
+            return Particle(argument[0]), Particle(argument[1])
+
+        if annotation in _basic_allowed_annotations and argument is None:
+            raise TypeError(f"{parameter} may not be None.")
+
+        particle = _physical_particle_factory(argument, Z=Z, mass_numb=mass_numb)
+
+        self.verify_charge_categorization(particle)
```

Comment:
I had been wondering about that, but tentatively decided that another layer of abstraction would make the code a bit harder to navigate, and it is only needed in this one place.  I'm inclined to keep it as is, at least for the time being, since there isn't a complicated control structure in this block.

Reply:
Ok, if this is the only place it occurs then agreed, no need to abstract it